xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -152,275 +188,185 @@ pmap_registry_search(xlator_t *this, const char *brickname,
-         * brick and get a disconnect and then immediately try to start
-         * the same brick instance based on another friend update
-         * request. And then if for the very first brick even if the
-         * process doesn't come up at the end sign in event gets sent
-         * and we end up having two duplicate portmap entries for the
-         * same brick. Since in brick start we mark the previous port as
-         * free, its better to consider a sign in request as no op if
-         * the corresponding port type is marked as free
-         */
-        goto out;
-    }
-    if (pmap->ports[p].brickname) {
-        char *tmp = pmap->ports[p].brickname;
-        asprintf(&pmap->ports[p].brickname, "%s %s", tmp, brickname);
-        free(tmp);
-    } else {
-        pmap->ports[p].brickname = strdup(brickname);
-    }
-    pmap->ports[p].type = type;
-    pmap->ports[p].xprt = xprt;
+    *new_port = tmp_port;
 
-    gf_msg("pmap", GF_LOG_INFO, 0, GD_MSG_BRICK_ADD,
-           "adding brick %s on port %d", brickname, port);
-
-    if (pmap->last_alloc < p)
-        pmap->last_alloc = p;
-out:
     return 0;
 }
 
+/* Add the port details to a list in pmap */
+
 int
-pmap_registry_extend(xlator_t *this, int port, const char *brickname)
+pmap_add_port_to_list(xlator_t *this, int port, char *brickname, void *xprt)
 {
     struct pmap_registry *pmap = NULL;
-    char *old_bn;
-    char *new_bn;
-    size_t bn_len;
-    char *entry;
-    int found = 0;
-
-    pmap = pmap_registry_get(this);
-
-    if (port > pmap->max_port) {
-        return -1;
-    }
+    struct pmap_ports *new_port = NULL;
+    int ret = -1;
 
-    switch (pmap->ports[port].type) {
-        case GF_PMAP_PORT_LEASED:
-        case GF_PMAP_PORT_BRICKSERVER:
-            break;
-        default:
-            return -1;
-    }
+    GF_ASSERT(this);
 
-    old_bn = pmap->ports[port].brickname;
-    if (old_bn) {
-        bn_len = strlen(brickname);
-        entry = strstr(old_bn, brickname);
-        while (entry) {
-            found = 1;
-            if ((entry != old_bn) && (entry[-1] != ' ')) {
-                found = 0;
-            }
-            if ((entry[bn_len] != ' ') && (entry[bn_len] != '\0')) {
-                found = 0;
-            }
-            if (found) {
-                return 0;
-            }
-            entry = strstr(entry + bn_len, brickname);
-        }
-        asprintf(&new_bn, "%s %s", old_bn, brickname);
-    } else {
-        new_bn = strdup(brickname);
-    }
+    pmap = pmap_registry_get(this);
 
-    if (!new_bn) {
-        return -1;
+    ret = pmap_port_new(this, port, brickname, xprt, &new_port);
+    if (ret) {
+        gf_msg(this->name, GF_LOG_ERROR, errno, GD_MSG_NO_MEMORY,
+               "Failed to allocate memory");
+        goto out;
     }
 
-    pmap->ports[port].brickname = new_bn;
-    free(old_bn);
+    cds_list_add(&new_port->port_list, &pmap->ports);
```

Comment:
Is this safe ? we don't use locks here so, unless callers ensure atomicity, we could have problems here.

Reply:
Yeah, the locking is taken care of by the mgmt framework before reaching this state.